go/worker/compute/executor/committee/proposals.go

```diff
@@ -0,0 +1,82 @@
+package committee
+
+import (
+	"fmt"
+
+	"github.com/google/btree"
+
+	p2pError "github.com/oasisprotocol/oasis-core/go/p2p/error"
+)
+
+// maxPendingProposals is the maximum number of pending proposals that can be queued.
+const maxPendingProposals = 16
+
+// pendingProposals is a priority queue of pending proposals, ordered by round.
+type pendingProposals struct {
+	q *btree.BTreeG[*unresolvedBatch]
+}
+
+func proposalLessFunc(a, b *unresolvedBatch) bool {
+	return a.proposal.Header.Round < b.proposal.Header.Round
+}
+
+func newPendingProposals() *pendingProposals {
+	return &pendingProposals{
+		q: btree.NewG[*unresolvedBatch](2, proposalLessFunc),
+	}
+}
+
+// addPendingProposalLocked adds a new pending proposal that MUST HAVE already undergone basic
+// validity checks and is therefore considered a valid proposal for the given round, but the node's
+// local consensus view may not yet be ready to process the proposal.
+//
+// Must be called with the n.commonNode.CrossNode lock held.
+func (n *Node) addPendingProposalLocked(batch *unresolvedBatch) error {
+	currentRound := n.commonNode.CurrentBlock.Header.Round
+	round := batch.proposal.Header.Round
+
+	// Drop any past proposals.
+	if round <= currentRound {
+		return p2pError.Permanent(fmt.Errorf("proposal round is in the past")) // Do not forward.
+	}
+
+	n.pendingProposals.q.ReplaceOrInsert(batch)
+
+	// In case of overflows, remove the round that is the most in the future.
+	if n.pendingProposals.q.Len() >= maxPendingProposals {
+		n.pendingProposals.q.DeleteMax()
+	}
+
+	return nil
+}
+
+// handlePendingProposalsLocked attempts to handle any pending proposals.
+//
+// Must be called with the n.commonNode.CrossNode lock held.
+func (n *Node) handlePendingProposalsLocked() {
+	if n.pendingProposals.q.Len() == 0 {
+		return
+	}
+
+	currentRound := n.commonNode.CurrentBlock.Header.Round
+
+	for {
+		batch, ok := n.pendingProposals.q.DeleteMin()
+		if !ok {
+			return
+		}
+
+		// Fast skip proposals that are no longer valid.
+		if batch.proposal.Header.Round <= currentRound {
+			continue
+		}
+
+		// Skip proposals that fail to be handled.
+		if err := n.handleProposalLocked(batch); err != nil {
```

Comment:
Currently this is not a problem, but in the future we can fall into a loop here as the batch is re-added to the queue. Btw, this error never happens.

Reply:
Yes currently the error never happens because we do the quick check above.

Reply:
I see, this should be removed and it doesn't make sense to ever handle more than one valid proposal (even if handling would fail).

Reply:
Fixed.